### PR TITLE
Convert pulsar 2.8 groovy test to Java

### PR DESCRIPTION
**Pull request by hadesy**
part of https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7195

### CODE REVIEW

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.PulsarContainer;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.utility.DockerImageName;
+
+public class PulsarClientTest {
```
Comment (jeanbisutti) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:50]:
```suggestion
class PulsarClientTest {
```

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+    brokerHost = pulsar.getHost();
+    brokerPort = pulsar.getMappedPort(6650);
+    client = PulsarClient.builder().serviceUrl(pulsar.getPulsarBrokerUrl()).build();
+    admin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build();
+  }
+
+  @AfterAll
+  static void afterAll() throws PulsarClientException {
+
```
Comment (jeanbisutti) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:90]:
```suggestion
```

**Review by jeanbisutti — COMMENTED** [review]
The `def "test consume non-partitioned topic using batchReceiveAsync"` test may be missing?

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+            .subscriptionName("test_sub")
+            .topic(topic)
+            .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+            .messageListener(
+                (MessageListener<String>)
+                    (consumer, msg) -> {
+                      try {
+                        consumer.acknowledge(msg);
+                      } catch (PulsarClientException e) {
+                        throw new AssertionError(e);
```
Comment (jeanbisutti) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:148]: It's not an error related to an assertion. The semantic of a previous code version with a `RuntimeException` seems more appropriate, or with an `IllegalStateException`. Same remark for other tests.

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+            .topic(topic)
+            .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+            .messageListener(
+                (MessageListener<String>)
+                    (consumer, msg) -> {
+                      try {
+                        consumer.acknowledge(msg);
+                      } catch (PulsarClientException e) {
+                        throw new RuntimeException(e);
+                      }
```
Comment (laurit) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:149]: I'd create a helper method for this as it is used in multiple tests

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+                        .hasAttributesSatisfyingExactly(
+                            receiveAttributes(topic, msgId.toString(), false)),
+                span ->
+                    span.hasName(topic + " process")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(2))
+                        .hasAttributesSatisfyingExactly(
+                            processAttributes(topic, msgId.toString(), false))));
+  }
+
```
Comment (laurit) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:184]: groovy test has an additional test named `test consume non-partitioned topic using receive` I could not find a matching test here

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+            .subscriptionName("test_sub")
+            .topic(topic)
+            .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+            .subscribe();
+    producer = client.newProducer(Schema.STRING).topic(topic).enableBatching(false).create();
+
+    String msg = "test";
+    MessageId msgId = testing.runWithSpan("parent", () -> producer.send(msg));
+
+    Message<String> receivedMsg = consumer.receive();
```
Comment (laurit) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:201]: although this test is named `ReceiveAsync` it uses `receive` not `receiveAsync` that was used in the groovy test

File: instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+    CompletableFuture<Messages<String>> result =
+        testing.runWithSpan(
+            "receive-parent",
+            () ->
+                consumer
+                    .batchReceiveAsync()
+                    .whenComplete(
+                        (messages, throwable) -> {
+                          if (throwable != null) {
+                            throw new RuntimeException(throwable);
```
Comment (laurit) [inline instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/PulsarClientTest.java:335]: throwing an exception from here doesn't make sense. I'd just ignore the `throwable` like the groovy test does or replace `whenComplete` with `thenApply`

**Review by laurit — APPROVED** [review]

**Pull request merged**